homeassistant/components/brunt/cover.py

```diff
@@ -1,92 +1,108 @@
 """Support for Brunt Blind Engine covers."""
 
+from datetime import timedelta
 import logging
 
-from brunt import BruntAPI
-import voluptuous as vol
+import async_timeout
+from brunt import BruntClientAsync, Thing
 
 from homeassistant.components.cover import (
     ATTR_POSITION,
     DEVICE_CLASS_WINDOW,
-    PLATFORM_SCHEMA,
     SUPPORT_CLOSE,
     SUPPORT_OPEN,
     SUPPORT_SET_POSITION,
     CoverEntity,
 )
-from homeassistant.const import ATTR_ATTRIBUTION, CONF_PASSWORD, CONF_USERNAME
-import homeassistant.helpers.config_validation as cv
+from homeassistant.const import ATTR_ATTRIBUTION
+from homeassistant.core import callback
+from homeassistant.helpers.update_coordinator import (
+    CoordinatorEntity,
+    DataUpdateCoordinator,
+    UpdateFailed,
+)
+
+from .const import (
+    ATTR_REQUEST_POSITION,
+    ATTRIBUTION,
+    CLOSED_POSITION,
+    DOMAIN,
+    OPEN_POSITION,
+)
 
 _LOGGER = logging.getLogger(__name__)
 
 COVER_FEATURES = SUPPORT_OPEN | SUPPORT_CLOSE | SUPPORT_SET_POSITION
-
-ATTR_REQUEST_POSITION = "request_position"
-NOTIFICATION_ID = "brunt_notification"
-NOTIFICATION_TITLE = "Brunt Cover Setup"
-ATTRIBUTION = "Based on an unofficial Brunt SDK."
-
-CLOSED_POSITION = 0
-OPEN_POSITION = 100
-
-PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
-    {vol.Required(CONF_USERNAME): cv.string, vol.Required(CONF_PASSWORD): cv.string}
-)
+REGULAR_INTERVAL = timedelta(seconds=20)
+FAST_INTERVAL = timedelta(seconds=1)
 
 
-def setup_platform(hass, config, add_entities, discovery_info=None):
+async def async_setup_entry(hass, entry, async_add_entities, discovery_info=None):
     """Set up the brunt platform."""
+    bapi: BruntClientAsync = hass.data[DOMAIN][entry.entry_id]
 
-    username = config[CONF_USERNAME]
-    password = config[CONF_PASSWORD]
-
-    bapi = BruntAPI(username=username, password=password)
-    try:
-        things = bapi.getThings()["things"]
-        if not things:
-            _LOGGER.error("No things present in account")
-        else:
-            add_entities(
-                [
-                    BruntDevice(bapi, thing["NAME"], thing["thingUri"])
-                    for thing in things
-                ],
-                True,
-            )
-    except (TypeError, KeyError, NameError, ValueError) as ex:
-        _LOGGER.error("%s", ex)
-        hass.components.persistent_notification.create(
-            "Error: {ex}<br />You will need to restart hass after fixing.",
-            title=NOTIFICATION_TITLE,
-            notification_id=NOTIFICATION_ID,
-        )
-
-
-class BruntDevice(CoverEntity):
+    async def async_update_data():
+        """Fetch data from the Brunt endpoint for all Things."""
+        try:
+            async with async_timeout.timeout(10):
+                things = await bapi.async_get_things(force=True)
+                return {thing.SERIAL: thing for thing in things}
+        except (TypeError, KeyError, NameError, ValueError) as err:
```

Comment:
If you would implement reauth (or at least catch it), it would be good to do it here as well. Otherwise, it will keep retrying until someone reboots Home Assistant. Since the DataUpdateCoordinator keeps polling, you could even get banned from the service, if they implement a brute force protection.

Reply:
Included the reauth and tests for it! thanks!